File: web/tests/core/errors/AppError.test.ts
import { describe, it, expect } from 'vitest';
import { AppError } from '@/core/errors';

describe('AppError Taxonomy', () => {
  it('should create INVALID_INPUT error with 400 status', () => {
    const error = AppError.invalidInput('Booking code format is invalid', { code: '123' });
    expect(error).toBeInstanceOf(AppError);
    expect(error).toBeInstanceOf(Error);
    expect(error.code).toBe('INVALID_INPUT');
    expect(error.statusCode).toBe(400);
    expect(error.message).toBe('Booking code format is invalid');
    expect(error.details).toEqual({ code: '123' });
  });

  it('should create BOOKING_CODE_NOT_FOUND error with 404 status', () => {
    const error = AppError.notFound();
    expect(error.code).toBe('BOOKING_CODE_NOT_FOUND');
    expect(error.statusCode).toBe(404);
    expect(error.message).toBe('The provided Betway booking code could not be found or has expired.');
    expect(error.details).toBeNull();
  });

  it('should create STALE_SELECTIONS error with 422 status', () => {
    const error = AppError.staleSelections('Match already started');
    expect(error.code).toBe('STALE_SELECTIONS');
    expect(error.statusCode).toBe(422);
    expect(error.message).toBe('Match already started');
  });

  it('should create UPSTREAM_BETWAY_ERROR error with 502 status', () => {
    const error = AppError.upstreamError('Betway gateway timed out');
    expect(error.code).toBe('UPSTREAM_BETWAY_ERROR');
    expect(error.statusCode).toBe(502);
    expect(error.message).toBe('Betway gateway timed out');
  });

  it('should create INTERNAL_SERVER_ERROR error with 500 status', () => {
    const error = AppError.internal();
    expect(error.code).toBe('INTERNAL_SERVER_ERROR');
    expect(error.statusCode).toBe(500);
    expect(error.message).toBe('An unexpected internal server error occurred.');
  });

  it('should serialize correctly to response envelope via toResponse()', () => {
    const error = AppError.notFound('Code not found', { bookingCode: 'INVALID' });
    const response = error.toResponse();

    expect(response).toEqual({
      success: false,
      error: {
        code: 'BOOKING_CODE_NOT_FOUND',
        message: 'Code not found',
        details: { bookingCode: 'INVALID' },
      },
    });
  });

  it('should wrap native errors using AppError.from()', () => {
    const nativeErr = new Error('Unexpected database failure');
    const wrapped = AppError.from(nativeErr);

    expect(wrapped).toBeInstanceOf(AppError);
    expect(wrapped.code).toBe('INTERNAL_SERVER_ERROR');
    expect(wrapped.statusCode).toBe(500);
    expect(wrapped.message).toBe('Unexpected database failure');
  });

  it('should return existing AppError instance untouched when passed to AppError.from()', () => {
    const existing = AppError.invalidInput('Bad format');
    const result = AppError.from(existing);

    expect(result).toBe(existing);
  });
});

File: mobile/test/domain/models/app_error_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/domain/models/app_error.dart';

void main() {
  group('AppError', () {
    test('creates invalidInput error with correct code and status 400', () {
      final error = AppError.invalidInput(
        message: 'Code format invalid',
        details: {'field': 'code'},
      );
      expect(error.code, AppErrorCodes.invalidInput);
      expect(error.message, 'Code format invalid');
      expect(error.statusCode, 400);
      expect(error.details, {'field': 'code'});
    });

    test('creates notFound error with correct code and status 404', () {
      final error = AppError.notFound();
      expect(error.code, AppErrorCodes.bookingCodeNotFound);
      expect(error.statusCode, 404);
      expect(error.message, contains('could not be found'));
    });

    test('creates staleSelections error with correct code and status 422', () {
      final error = AppError.staleSelections();
      expect(error.code, AppErrorCodes.staleSelections);
      expect(error.statusCode, 422);
      expect(error.message, contains('no longer active'));
    });

    test('creates upstreamError with correct code and status 502', () {
      final error = AppError.upstreamError();
      expect(error.code, AppErrorCodes.upstreamBetwayError);
      expect(error.statusCode, 502);
      expect(error.message, contains('upstream service'));
    });

    test('creates networkError with correct code', () {
      final error = AppError.networkError();
      expect(error.code, AppErrorCodes.networkError);
      expect(error.statusCode, isNull);
      expect(error.message, contains('internet connection'));
    });

    test('creates internal error with correct code and status 500', () {
      final error = AppError.internal();
      expect(error.code, AppErrorCodes.internalServerError);
      expect(error.statusCode, 500);
      expect(error.message, contains('unexpected internal error'));
    });

    test('implements value equality via Equatable', () {
      final err1 = AppError.notFound(message: 'Slip not found');
      final err2 = AppError.notFound(message: 'Slip not found');
      final err3 = AppError.invalidInput(message: 'Slip not found');

      expect(err1, equals(err2));
      expect(err1.hashCode, equals(err2.hashCode));
      expect(err1, isNot(equals(err3)));
    });

    test('toString returns formatted string representation', () {
      final error = AppError.notFound(message: 'Not found');
      expect(
        error.toString(),
        'AppError(code: BOOKING_CODE_NOT_FOUND, message: Not found, statusCode: 404)',
      );
    });
  });
}
